napari/layers/_multiscale_data.py

```diff
@@ -0,0 +1,121 @@
+    max_size : Sequence[int], optional
+        Maximum size of a displayed tile in pixels, by default`data[-1].shape`
+
+    Raises
+    ------
+    ValueError
+        If `data` is empty or is not a list, tuple, or ndarray.
+    TypeError
+        If any of the items in `data` don't provide `LayerDataProtocol`.
+    """
+
+    def __init__(
+        self,
+        data: Sequence[LayerDataProtocol],
+        max_size: Optional[Sequence[int]] = None,
+    ) -> None:
+        self._data: List[LayerDataProtocol] = list(data)
+        if not self._data:
+            raise ValueError("Multiscale data must be a (non-empty) sequence")
+        for d in self._data:
+            assert_protocol(d)
+
+        self.max_size = self._data[-1].shape if max_size is None else max_size
+        self.downsample_factors = (
+            np.array([d.shape for d in data]) / data[0].shape
+        )
+
+    @property
+    def dtype(self) -> np.dtype:
+        """Return dtype of the first scale.."""
+        return self._data[0].dtype
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Shape of multiscale is just the biggest shape."""
+        return self._data[0].shape
+
+    @property
+    def shapes(self) -> Tuple[Tuple[int, ...], ...]:
+        """Tuple shapes for all scales."""
+        return tuple(im.shape for im in self._data)
+
+    def __getitem__(  # type: ignore [override]
+        self, key: Union[int, Tuple[slice, ...]]
+    ) -> LayerDataProtocol:
+        """Multiscale indexing.
+
+        When indexing with a single integer, the behavior is exactly like the
+        input squence (it retrieves the sepcified level).
+
+        When indexing with a tuple of slices, it will extract a chunk from the
+        appropriate level using `compute_multiscale_level_and_corners`.
+
+        All other `key` types are currently undefined and raise a
+        NotImplementedError (for mixed tuple of slice and int), or TypeError
+        (for everything else.)
+        """
+        if isinstance(key, int):
+            return self._data[key]
+
+        if isinstance(key, tuple):
+            if all(isinstance(idx, slice) for idx in key):
+                corners = np.array(
+                    [
+                        (sl.start or 0, sl.stop or size)
+                        for sl, size in zip(key, self._data[0].shape)
+                    ]
+                )
+                level, corners = compute_multiscale_level_and_corners(
+                    corners, self.max_size, self.downsample_factors
+                )
+                return self._data[level][corners]
+            else:
+                raise NotImplementedError("cannot handle slices and ints")
+
+        raise TypeError(f"Cannot index multiscale data with {type(key)}")
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    def __eq__(self, other) -> bool:
+        return self._data == other
+
+    def __add__(self, other) -> bool:
```

Comment:
Do we need to define `__add__`, `__mul__`, and `__rmul__`? If so, why? And should they be in the `LayerDataProtocol`? I don't think the current implementations for `__mul__` and `__rmul__` are going to work if `self._data` is a `list`, but maybe I'm missing something. `__add__` will, but I'm not sure if it's doing what we want.

Reply:
> Do we need to define __add__, __mul__, and __rmul__? If so, why?

we don't really _need_ them.  I only did that to preserve the `list` behavior from the current "API" (wherein multiscale data is a plain list of arrays).  My goal at the time was to make this wrapper as close as possible to the input list it was accepting, so alleviate concerns that we were creating a custom object that didn't behavior like the current API.  This object can be seen as a (non-mutable) sequence of arrays that just adds `dtype` and `shape` ... and now with @jni's improvements to `__getitem__`  performs some sane defaults when you index with slices.

> And should they be in the `LayerDataProtocol`

Nope, those dunder methods are likely reserved for array-like element-wise behavior on the main protocol... whereas here they're more for the outer sequence.  e.g. `LayerDataProtocol.__add__` would likely be like `np.add` _not_ `list.__add__`

> I don't think the current implementations for __mul__ and __rmul__ are going to work if self._data is a list, but maybe I'm missing something

Reply:
> I don't think the current implementations for __mul__ and __rmul__ are going to work if self._data is a list, but maybe I'm missing something

they are simply there to pass on the behavior to the list, which is what we want (this isn't an array, it's a list).  Not sure if this was clear, but the point here isn't to make a complex multi-scale object that turns the inner list into a fully capable arraylike.  It's just a first-step to make sure that all of the things that query `layer.data.shape` and `layer.data.dtype` don't fully crash, and in every other way behaves like the current API of a list of arrays.

Reply:
Well, actually it's declared as a `Sequence`, which I don't think expects these methods (though certainly won't reject them). My guess is that we use the extra dunder methods like `list.__add__` in our code (where before we had a plain list), so you added these to make that usage continue to work, which seems reasonable.

If that's the case, my next question is should we make `MultiScaleData` extend `list` instead of `Sequence`, but my guess is that we don't want to commit to inheriting/implementing everything expected from a list. Is that right?

Reply:
> my next question is should we make MultiScaleData extend list instead of Sequence, but my guess is that we don't want to commit to inheriting/implementing everything expected from a list. Is that right?

I didn't want it to be mutable.  so, i did want to declare it as a `Sequence`, but provide a few of the things that one could already do with `layer.data` when it was a pure list.